lib/ts/xmlhttprequest.ts

```diff
@@ -0,0 +1,429 @@
+                    // .origin gives the port as well..
+                    doNotDoInterception = !shouldDoInterceptionBasedOnUrl(
+                        WindowHandlerReference.getReferenceOrThrow().windowHandler.location.getOrigin(),
+                        AuthHttpRequestFetch.config.apiDomain,
+                        AuthHttpRequestFetch.config.cookieDomain
+                    );
+                } else {
+                    throw err;
+                }
+            }
+
+            logDebugMessage("send: Value of doNotDoInterception: " + doNotDoInterception);
+            if (doNotDoInterception) {
+                logDebugMessage("send: Returning without interception");
+                return actual.send(body);
+            }
+            logDebugMessage("send: Interception started");
+
+            ProcessState.getInstance().addState(PROCESS_STATE.CALLING_INTERCEPTION_REQUEST);
+
+            (async function() {
+                preRequestIdToken = await getIdRefreshToken(true);
+
+                if (preRequestIdToken.status === "EXISTS") {
+                    const antiCsrfToken = await AntiCsrfToken.getToken(preRequestIdToken.token);
+                    if (antiCsrfToken !== undefined) {
+                        logDebugMessage("send: Adding anti-csrf token to request");
+                        actual.setRequestHeader("anti-csrf", antiCsrfToken);
+                    }
+                }
+
+                if (AuthHttpRequestFetch.config.autoAddCredentials) {
+                    logDebugMessage("send: Adding credentials include");
+                    self.withCredentials = true;
+                }
+
+                if (
+                    !requestHeaders.some(i => {
+                        i.name === "rid";
+                    })
+                ) {
+                    logDebugMessage("send: Adding rid header: anti-csrf");
+                    actual.setRequestHeader("rid", "anti-csrf");
+                } else {
+                    logDebugMessage("send: rid header was already there in request");
+                }
+
+                logDebugMessage("send: Making user's http call");
+                return actual.send(body);
+            })();
+        };
+
+        self.setRequestHeader = function(name: string, value: string) {
+            listOfFunctionCallsInProxy.push((xhr: XMLHttpRequestType) => {
+                xhr.setRequestHeader(name, value);
+            });
+            // TODO: If this is called twice on the same key, is the older version
+            // removed or is the newer value just appended..
+            if (
+                requestHeaders.some(i => {
+                    return i.name === name;
+                })
+            ) {
+                requestHeaders = requestHeaders.filter(i => {
+                    return i.name !== name;
+                });
+            }
+            requestHeaders.push({ name, value });
+            actual.setRequestHeader(name, value);
+        };
+
```

Comment:
How to handle progress events? Cause the actual instance will say progress is finished when it's done, but then if we retry, then progress will start again.